runtime/runtime/src/actions.rs

```diff
@@ -789,6 +848,73 @@ pub(crate) fn action_add_key(
     Ok(())
 }
 
+pub(crate) fn action_add_gas_key(
+    apply_state: &ApplyState,
+    state_update: &mut TrieUpdate,
+    account: &mut Account,
+    result: &mut ActionResult,
+    account_id: &AccountId,
+    add_gas_key: &AddGasKeyAction,
+) -> Result<(), StorageError> {
+    if get_gas_key(state_update, account_id, &add_gas_key.public_key)?.is_some() {
+        result.result = Err(ActionErrorKind::GasKeyAlreadyExists {
+            account_id: account_id.to_owned(),
+            public_key: add_gas_key.public_key.clone().into(),
+        }
+        .into());
+        return Ok(());
+    }
+    let gas_key = GasKey {
+        num_nonces: add_gas_key.num_nonces,
+        permission: add_gas_key.permission.clone(),
+        balance: Balance::ZERO,
+    };
+    set_gas_key(state_update, account_id.clone(), add_gas_key.public_key.clone(), &gas_key);
+
+    for i in 0..gas_key.num_nonces {
+        let nonce = (apply_state.block_height - 1)
+            * near_primitives::account::AccessKey::ACCESS_KEY_NONCE_RANGE_MULTIPLIER;
+        set_gas_key_nonce(
+            state_update,
+            account_id.clone(),
+            add_gas_key.public_key.clone(),
+            i,
+            nonce,
+        );
+    }
+    account.set_storage_usage(
+        account
+            .storage_usage()
+            .checked_add(gas_key_storage_cost(
+                &apply_state.config.fees,
+                &add_gas_key.public_key,
+                &gas_key,
+            ))
+            .ok_or_else(|| {
+                StorageError::StorageInconsistentState(format!(
+                    "Storage usage integer overflow for account {}",
+                    account_id
+                ))
+            })?,
+    );
+    Ok(())
+}
+
+fn gas_key_storage_cost(
+    fee_config: &RuntimeFeesConfig,
+    public_key: &PublicKey,
+    gas_key: &GasKey,
+) -> StorageUsage {
+    let storage_config = &fee_config.storage_usage_config;
+    let nonce_storage_usage = gas_key.num_nonces as u64
+        * (borsh::object_length(&0u64).unwrap() as u64 + storage_config.num_extra_bytes_record);
```

Comment:
So this accounts for the nonce itself; do we also need to account for nonce index that is part of the key?

Reply:
I am not sure about repeated public key. I don't actually know how we implement the trie, but I would expect that common key prefixes to be accounted for only once. So because nonce gas keys will share prefix with gas keys and have the same prefix, I would expect that prefix to be in a single node in trie. This would be different with indices, since each nonce key would have a different index.